wolfcrypt/src/port/xilinx/xil-versal-trng.c

```diff
@@ -0,0 +1,195 @@
+{
+    s32 ret = XTrngpsv_RunKAT(&trng);
+    if (ret == XTRNGPSV_SUCCESS)
+        ret = XTrngpsv_RunHealthTest(&trng);
+    return ret;
+}
+
+#if !defined(HAVE_HASHDRBG)
+static WC_INLINE void array_add_one(byte* data, word32 dataSz)
+{
+    int i;
+
+    for (i = dataSz - 1; i >= 0; i--)
+    {
+        data[i]++;
+        if (data[i] != 0) break;
+    }
+}
+
+static WC_INLINE void usercfg_add_nonce(XTrngpsv_UsrCfg* user_cfg, byte* nonce, word32 nonceSz)
+{
+    word32 nonce_written = 0;
+    byte *pers_string = (byte*)user_cfg->PersString;
+    byte *nonce_rd = nonce;
+    do {
+        word32 to_write = nonce_written + nonceSz <= sizeof(user_cfg->PersString) ?
+                            nonceSz : sizeof(user_cfg->PersString) - nonce_written;
+        XMEMCPY(pers_string, nonce_rd, to_write);
+        if (nonce != nonce_rd)
+            array_add_one(pers_string, to_write);
+        nonce_rd = pers_string;
+        pers_string += to_write;
+        nonce_written += to_write;
+    } while (nonce_written < sizeof(user_cfg->PersString));
+    user_cfg->PersStrPresent = XTRNGPSV_TRUE;
+}
+#endif
+
+int wc_versal_trng_init(byte* nonce, word32 nonceSz)
+{
+#if !defined(HAVE_HASHDRBG)
+    /* Use the HRNG mode (with DF) which lets us extract FIPS compliant
+     * random data from the TRNG.
+     */
+    XTrngpsv_UsrCfg user_cfg = {
+            .Mode = XTRNGPSV_HRNG,
+            .SeedLife = WC_RESEED_INTERVAL,
+            .PredResistanceEn = XTRNGPSV_FALSE,
+            .DFDisable = XTRNGPSV_FALSE,
+            .DFLenMul = WC_XILINX_TRNG_DF_LEN_MUL,
+            .InitSeedPresent = XTRNGPSV_FALSE,
+            .PersStrPresent = XTRNGPSV_FALSE
+    };
+#else
+    /* Use the PTRNG mode (without DF) which lets us extract raw entropy
+     * to be used as a seed to the internal DRBG.
+     */
+    const XTrngpsv_UsrCfg user_cfg = {
+            .Mode = XTRNGPSV_PTRNG,
+            .SeedLife = 0U,
+            .PredResistanceEn = XTRNGPSV_FALSE,
+            .DFDisable = XTRNGPSV_TRUE,
+            .DFLenMul = 0U,
+            .InitSeedPresent =  XTRNGPSV_FALSE,
+            .PersStrPresent = XTRNGPSV_FALSE
+    };
+#endif
+    int ret = WC_HW_E;
+    XTrngpsv_Config *cfg;
+    s32 xret = 0;
+    if (trng.State == XTRNGPSV_HEALTHY) {
+        return 0;
+    }
+    cfg = XTrngpsv_LookupConfig(WOLFSSL_PSV_TRNG_DEV_ID);
+    if (!cfg) {
+        WOLFSSL_MSG("Could not lookup TRNG config");
+        goto out;
+    }
+    xret = XTrngpsv_CfgInitialize(&trng, cfg, cfg->BaseAddress);
+    if (xret)
+        goto out;
+    xret = versal_trng_selftest();
+    if (xret)
+        goto out;
+#if !defined(HAVE_HASHDRBG)
+    if (nonce)
+        usercfg_add_nonce(&user_cfg, nonce, nonceSz);
+#endif
+    xret = XTrngpsv_Instantiate(&trng, &user_cfg);
+    if (xret)
+        goto out;
+
+    ret = 0;
+
+out:
+    if (xret) {
+        WOLFSSL_MSG_EX("Xilinx API returned 0x%08x", xret);
+    }
+    return ret;
+}
+
+int wc_versal_trng_reset(void)
```

Comment:
Lets add comments above each of these functions, it does not have to be elaborate. At the minimum the success return value expected and a short one line brief.